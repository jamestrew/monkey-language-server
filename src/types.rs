#![allow(dead_code)]
use std::ops::{Deref, Range};

#[derive(PartialEq, Default)]
pub struct Spanned<T> {
    pub start: Position,
    pub end: Position,
    pub span: Range<usize>,
    data: T,
}

impl<T> Spanned<T> {
    pub fn new(start: Position, end: Position, span: Range<usize>, data: T) -> Self {
        Self {
            start,
            end,
            span,
            data,
        }
    }

    pub fn map<S>(&self, new_data: S) -> Spanned<S> {
        Spanned {
            start: self.start,
            end: self.end,
            span: self.span(),
            data: new_data,
        }
    }

    pub fn span(&self) -> Range<usize> {
        self.span.to_owned()
    }

    pub fn start(&self) -> usize {
        self.span.start
    }

    pub fn end(&self) -> usize {
        self.span.end
    }

    pub fn take(self) -> T {
        self.data
    }

    pub fn pos_rng_str(&self) -> String {
        format!("{:?}->{:?}", self.start, self.end)
}
}

impl<T> Deref for Spanned<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.data
    }
}

#[derive(Clone, Copy, PartialEq, Default)]
pub struct Position {
    row: usize,
    col: usize,
}

impl Position {
    pub fn new(row: usize, col: usize) -> Self {
        Self { row, col }
    }
}

impl std::fmt::Debug for Position {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "({},{})", self.row, self.col)
    }
}
