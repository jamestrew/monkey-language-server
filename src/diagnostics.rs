use thiserror::Error;
use tower_lsp::lsp_types::DiagnosticSeverity;

use crate::ast::Operator;
use crate::eval::Object;
use crate::types::Spanned;

pub type SpannedDiagnostic = Spanned<Diagnostics>;
pub type SpannedError = Spanned<MonkeyError>;
pub type SpannedWarning = Spanned<MonkeyWarning>;

#[derive(Debug, Clone, PartialEq)]
pub enum Diagnostics {
    Error(MonkeyError),
    Warning(MonkeyWarning),
    // Information,
    // Hint,
}

impl Diagnostics {
    pub fn severity(&self) -> DiagnosticSeverity {
        match self {
            Diagnostics::Error(_) => DiagnosticSeverity::ERROR,
            Diagnostics::Warning(_) => DiagnosticSeverity::WARNING,
        }
    }
}

impl std::fmt::Display for Diagnostics {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Diagnostics::Error(err) => write!(f, "{err}"),
            Diagnostics::Warning(warn) => write!(f, "{warn}"),
        }
    }
}

impl std::fmt::Debug for SpannedDiagnostic {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let msg = match &**self {
            Diagnostics::Error(err) => format!("{:?}", err),
            Diagnostics::Warning(warn) => format!("{:?}", warn),
        };
        write!(f, "{:?}({msg}, {})", self.severity(), self.pos_rng_str())
    }
}

#[derive(Error, Debug, Clone, PartialEq)]
pub enum MonkeyError {
    #[error("SyntaxError: Unexpected '{0}'.")]
    UnexpectedToken(String),

    #[error("SyntaxError: Unexpected eof.")]
    UnexpectedEof,

    #[error("SyntaxError: Expected '{0}' not found.")]
    ExpectedTokenNotFound(String),

    #[error("NameError: '{0}' is not defined.")]
    UnknownIdentifier(String),

    #[error("TypeError: bad operand type for unary +: '{0}'")]
    BadPrefixType(&'static str),

    #[error("TypeError: <{0}> {2} <{1}>")]
    UnknownOperator(&'static str, &'static str, &'static str),

    #[error("TypeError: object of type '{0}' is not callable")]
    BadFunctionCall(&'static str),
}

impl MonkeyError {
    pub fn new_unknown_op<T>(
        span: &Spanned<T>,
        left_obj: &Object,
        right_obj: &Object,
        op: Operator,
    ) -> SpannedError {
        span.map(Self::UnknownOperator(
            left_obj.typename(),
            right_obj.typename(),
            op.as_str(),
        ))
    }
}

impl std::fmt::Debug for SpannedError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let err = match &**self {
            MonkeyError::UnexpectedToken(_) => format!("UnexpectedToken(\"{}\")", **self),
            MonkeyError::UnexpectedEof => format!("UnexpectedEof(\"{}\")", **self),
            MonkeyError::ExpectedTokenNotFound(_) => {
                format!("ExpectedTokenNotFound(\"{}\")", **self)
            }
            MonkeyError::UnknownIdentifier(_) => format!("UnknownIdentifier(\"{}\")", **self),
            MonkeyError::BadPrefixType(_) => format!("BadPrefixType(\"{}\")", **self),
            MonkeyError::UnknownOperator(..) => format!("UnknownOperator(\"{}\")", **self),
            MonkeyError::BadFunctionCall(..) => format!("BadFunctionCall(\"{}\")", **self),
        };
        write!(f, "Err({err}, {})", self.pos_rng_str())
    }
}

impl From<MonkeyError> for Diagnostics {
    fn from(err: MonkeyError) -> Self {
        Diagnostics::Error(err)
    }
}

impl From<SpannedError> for SpannedDiagnostic {
    fn from(value: SpannedError) -> Self {
        value.transform()
    }
}

#[derive(Error, Debug, Clone, PartialEq)]
pub enum MonkeyWarning {
    #[error("Expression value is unused")]
    UnusedExpression,

    #[error("Code is unreachable")]
    UnreachableCode
}

impl std::fmt::Debug for SpannedWarning {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let err = match &**self {
            MonkeyWarning::UnusedExpression => format!("UnusedExpression(\"{}\")", **self),
            MonkeyWarning::UnreachableCode => format!("UnreachableCode(\"{}\")", **self),
        };
        write!(f, "Warn({err}, {})", self.pos_rng_str())
    }
}

impl From<MonkeyWarning> for Diagnostics {
    fn from(err: MonkeyWarning) -> Self {
        Diagnostics::Warning(err)
    }
}

impl From<SpannedWarning> for SpannedDiagnostic {
    fn from(value: SpannedWarning) -> Self {
        value.transform()
    }
}
