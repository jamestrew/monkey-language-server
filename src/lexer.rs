use std::ops::Range;

use logos::Logos;

use crate::diagnostics::{MonkeyError, SpannedError};
use crate::parser::TokenProvider;
use crate::types::{Position, Spanned};

#[derive(Logos, Debug, PartialEq, Clone, Copy)]
#[logos(skip r"[ \t\f]")]
pub enum TokenKind {
    #[token("=")]
    Assign,
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Asterisk,
    #[token("/")]
    ForwardSlash,
    #[token("!")]
    Bang,

    #[token("==")]
    Equal,
    #[token("!=")]
    NotEqual,
    #[token("<")]
    LT,
    #[token(">")]
    GT,

    #[token(",")]
    Comma,
    #[token(";")]
    Semicolon,
    #[token(":")]
    Colon,

    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token("[")]
    LBracket,
    #[token("]")]
    RBracket,

    // keywords
    #[token("let")]
    Let,
    #[token("fn")]
    Function,
    #[token("if")]
    If,
    #[token("else")]
    Else,
    #[token("true")]
    True,
    #[token("false")]
    False,
    #[token("return")]
    Return,
    #[token("nil")]
    Nil,

    #[regex("[a-zA-Z][a-zA-Z0-9_]*")]
    Identifier,
    #[regex("[0-9]+")]
    Int,
    #[regex(r#""(?:[^"]|\\")*""#)]
    Str,

    #[regex("\n|\r\n")]
    NewLine,
}

impl AsRef<TokenKind> for TokenKind {
    fn as_ref(&self) -> &TokenKind {
        self
    }
}

impl std::fmt::Display for TokenKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TokenKind::Assign => write!(f, "="),
            TokenKind::Plus => write!(f, "+"),
            TokenKind::Minus => write!(f, "-"),
            TokenKind::Asterisk => write!(f, "*"),
            TokenKind::ForwardSlash => write!(f, "/"),
            TokenKind::Bang => write!(f, "!"),
            TokenKind::Equal => write!(f, "=="),
            TokenKind::NotEqual => write!(f, "!="),
            TokenKind::LT => write!(f, "<"),
            TokenKind::GT => write!(f, ">"),
            TokenKind::Comma => write!(f, ","),
            TokenKind::Semicolon => write!(f, ";"),
            TokenKind::Colon => write!(f, ":"),
            TokenKind::LParen => write!(f, "("),
            TokenKind::RParen => write!(f, ")"),
            TokenKind::LBrace => write!(f, "{{"),
            TokenKind::RBrace => write!(f, "}}"),
            TokenKind::LBracket => write!(f, "["),
            TokenKind::RBracket => write!(f, "]"),
            TokenKind::Let => write!(f, "let"),
            TokenKind::Function => write!(f, "fn"),
            TokenKind::If => write!(f, "if"),
            TokenKind::Else => write!(f, "else"),
            TokenKind::True => write!(f, "true"),
            TokenKind::False => write!(f, "false"),
            TokenKind::Return => write!(f, "return"),
            TokenKind::Nil => write!(f, "nil"),
            TokenKind::Identifier => write!(f, "<identifier>"),
            TokenKind::Int => write!(f, "<int>"),
            TokenKind::Str => write!(f, "<str>"),
            TokenKind::NewLine => write!(f, r"\n"),
        }
    }
}

#[derive(PartialEq)]
pub struct _Token<'source> {
    pub kind: TokenKind,
    pub slice: &'source str,
}

impl<'source> _Token<'source> {
    #[cfg(test)]
    pub fn new(kind: TokenKind, slice: &'source str) -> _Token<'source> {
        Self { kind, slice }
    }

    pub fn from_lexer(lexer: &Lexer<'source>, kind: TokenKind) -> Token<'source> {
        let slice = lexer.current_slice();
        lexer.new_spanned_item(Self { kind, slice })
    }
}

pub type Token<'source> = Spanned<_Token<'source>>;
pub type TokenResult<'source> = Result<Token<'source>, SpannedError>;

pub fn token_result_span<T>(token_res: &TokenResult<'_>, data: T) -> Spanned<T> {
    match token_res {
        Ok(token) => token.map(data),
        Err(err) => err.map(data),
    }
}

pub struct Lexer<'source> {
    lexer: logos::Lexer<'source, TokenKind>,
    row: usize,
    last_newline_pos: usize,
}

impl<'source> Lexer<'source> {
    pub fn new(source: &'source str) -> Self {
        let lexer = TokenKind::lexer(source);

        Self {
            lexer,
            row: 0,
            last_newline_pos: 0,
        }
    }

    fn new_spanned_item<T>(&self, data: T) -> Spanned<T> {
        let span = self.lexer.span();
        let col_start = span.start - self.last_newline_pos;
        let col_end = span.end - self.last_newline_pos;

        let start = Position::new(self.row, col_start);
        let end = Position::new(self.row, col_end);

        Spanned::new(start, end, self.current_span(), data)
    }

    pub fn current_span(&self) -> Range<usize> {
        self.lexer.span().clone()
    }

    fn current_slice(&self) -> &'source str {
        self.lexer.slice()
    }
}

impl<'source> std::fmt::Debug for Token<'source> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Token({:#?}, {:#?}, {})",
            self.kind,
            self.slice,
            self.pos_rng_str()
        )
    }
}

impl<'source> Iterator for Lexer<'source> {
    type Item = TokenResult<'source>;

    fn next(&mut self) -> Option<Self::Item> {
        while let Some(token_res) = self.lexer.next() {
            match token_res {
                Ok(TokenKind::NewLine) => {
                    self.row += 1;
                    self.last_newline_pos = self.lexer.span().end;
                    continue;
                }
                Ok(token) => return Some(Ok(_Token::from_lexer(self, token))),
                Err(_) => {
                    let err = self
                        .new_spanned_item(MonkeyError::UnexpectedToken(self.lexer.slice().into()));
                    return Some(Err(err));
                }
            }
        }
        None
    }
}

impl<'source> TokenProvider<'source> for Lexer<'source> {
    fn next(&mut self) -> Option<TokenResult<'source>> {
        Iterator::next(self)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    enum T<'source> {
        Token(Token<'source>),
        Error(SpannedError),
    }

    impl<'source> std::fmt::Debug for T<'source> {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            match self {
                Self::Token(t) => write!(f, "{:?}", t),
                Self::Error(err) => write!(f, "{:?}", err),
            }
        }
    }

    fn debug_print(token_results: Vec<Result<Token<'_>, SpannedError>>) -> Vec<T> {
        let mut ret = Vec::new();
        for result in token_results {
            match result {
                Ok(token) => ret.push(T::Token(token)),
                Err(err) => ret.push(T::Error(err)),
            }
        }
        ret
    }

    macro_rules! snapshot {
        ($name:tt, $input:expr) => {
            #[test]
            fn $name() {
                let lexer = Lexer::new($input);
                let tokens: Vec<_> = lexer.collect();

                insta::with_settings!({
                    description => $input,
                }, {
                    insta::assert_debug_snapshot!(debug_print(tokens));
                })
            }
        };
    }

    snapshot!(check_span, "+\n+\n   +");
    snapshot!(simple_error, "@");
    snapshot!(simple_symbol, "+");
    snapshot!(more_symbols_1, "= + - *  / !");
    snapshot!(more_symbols_2, "==\n!=\n\t<>,;:");
    snapshot!(more_symbols_3, "(){}[]");
    snapshot!(keywords, "let\nfn\nif\nelse\ntrue\nfalse\nreturn nil");
    snapshot!(identifiers, "x\nabc");
    snapshot!(ints, "123");
    snapshot!(str, "\"hello world 123\"");

    snapshot!(assignments1, "let five = 5;");
    snapshot!(assignments2, "let ten = 10;");
    snapshot!(
        assignments3,
        r"let add = fn(x, y) {
            x + y;
        };"
    );
    snapshot!(assignments4, "let result = add(five, ten);");

    snapshot!(operators, "!-/*5;");
    snapshot!(equality1, "5 < 10 > 5;");
    snapshot!(equality2, "10 == 10;");
    snapshot!(equality3, "10 != 9;");
    snapshot!(
        conditional,
        r"if (5 < 10) {
            return true;
        } else {
            return false;
        }"
    );

    snapshot!(arrays, "[1, 2]");
    snapshot!(dictionary, r#"{ "foo": "bar" };"#);
}
