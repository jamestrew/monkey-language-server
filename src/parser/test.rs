use std::ops::Range;

use paste::paste;

use super::*;
use crate::diagnostics::SpannedDiagnostic;
use crate::lexer::*;
use crate::test_util::input_diagnostics;
use crate::types::*;

pub fn debug_new(
    start: Position,
    end: Position,
    span: Range<usize>,
    kind: TokenKind,
    slice: &'_ str,
) -> Token<'_> {
    Spanned::new(start, end, span, _Token::new(kind, slice))
}

#[test]
fn new_parser() {
    let source = "let a = 123;";
    let parser = Parser::from_source(source);

    let first = debug_new(
        Position::new(0, 0),
        Position::new(0, 3),
        0..3,
        TokenKind::Let,
        "let",
    );
    let second = debug_new(
        Position::new(0, 4),
        Position::new(0, 5),
        4..5,
        TokenKind::Identifier,
        "a",
    );

    assert_eq!(parser.curr_token, Some(Ok(first)));
    assert_eq!(parser.peek_token, Some(Ok(second)));
}

macro_rules! debug_snapshot {
    ($name:ident, $input:expr) => {
        paste! {
            #[test]
            fn $name() {
                let program = Parser::from_source($input).parse_program();
                let errors = program
                    .collect_errors()
                    .into_iter()
                    .map(|err| err.into())
                    .collect::<Vec<SpannedDiagnostic>>();
                let errors = input_diagnostics($input, errors);
                let result = format!("{:#?}\n\n===================================\n\n{}", program, errors);
                insta::with_settings!({
                    description => $input,
                }, {
                    insta::assert_snapshot!(result);
                });
            }
        }
    };
}

debug_snapshot!(lexer_erro, "@");
debug_snapshot!(bool_expr, "true");

// maybe these are warnings? "Expression value is unused" <- python
debug_snapshot!(int_expr, "123");
debug_snapshot!(nil_expr, "nil");
debug_snapshot!(string_literal, "\"hello world\"");
debug_snapshot!(identifier, "x");

debug_snapshot!(let_statement_1, "let x = 1;");
debug_snapshot!(let_statement_2, "let x = 2;;;");
debug_snapshot!(let_statement_broken_1, "let x = ;");
debug_snapshot!(let_statement_broken_2, "let x;");
debug_snapshot!(let_statement_broken_3, "let;");
debug_snapshot!(let_statement_broken_4, "le a = 1;");
debug_snapshot!(let_statement_broken_5, "let a = @;");
debug_snapshot!(let_statement_broken_6, "let a = @;42;");
debug_snapshot!(let_statement_broken_7, "let a = 1 2;");

debug_snapshot!(return_happy_1, "return 1;");
debug_snapshot!(return_happy_2, "return nil;");
debug_snapshot!(return_happy_3, "return;");
debug_snapshot!(return_happy_4, "return 1 + 1;");

debug_snapshot!(return_unhappy_1, "return @;");
debug_snapshot!(return_unhappy_2, "return ];");
debug_snapshot!(return_unhappy_3, "return [;");
debug_snapshot!(return_unhappy_4, "return x y;");

debug_snapshot!(prefix_happy_1, "-1");
debug_snapshot!(prefix_happy_2, "!true");
debug_snapshot!(prefix_happy_3, "let b = !true");

debug_snapshot!(prefix_unhappy_1, "-@");

debug_snapshot!(infix_happy_1, "1 + 1;\na - 2;\n1 * 1;\n1 / 2;");
debug_snapshot!(
    infix_happy_2,
    "\"foo\" == a;\n1 == 1;\n2 != 1;\n4 < 5;\n 5 > 4;"
);

// NEEDS IMPROVEMENT
// error position is on the operator
debug_snapshot!(infix_unhappy_1, "1 + +;");
debug_snapshot!(infix_unhappy_2, "1 +;");
debug_snapshot!(infix_unhappy_3, "1 + /");
debug_snapshot!(infix_unhappy_4, "1 + @");
debug_snapshot!(infix_unhappy_5, "@ + 1");

debug_snapshot!(grouped_expr, "(1 + 1)");

debug_snapshot!(operator_precedence_1, "-a * b");
debug_snapshot!(operator_precedence_2, "!-a");
debug_snapshot!(operator_precedence_3, "a + b + c");
debug_snapshot!(operator_precedence_4, "a + b - c");
debug_snapshot!(operator_precedence_5, "a * b * c");
debug_snapshot!(operator_precedence_6, "a * b / c");
debug_snapshot!(operator_precedence_7, "a + b / c");
debug_snapshot!(operator_precedence_8, "a / b + c");
debug_snapshot!(operator_precedence_9, "a + b * c + d / e - f");
debug_snapshot!(operator_precedence_10, "5 > 4 == 3 < 4");
debug_snapshot!(operator_precedence_11, "5 < 4 != 3 > 4");
debug_snapshot!(operator_precedence_12, "3 + 4 * 5 == 3 * 1 + 4 * 5");
debug_snapshot!(operator_precedence_13, "true");
debug_snapshot!(operator_precedence_14, "false");
debug_snapshot!(operator_precedence_15, "3 > 5 == false");
debug_snapshot!(operator_precedence_16, "3 < 5 == true");
debug_snapshot!(operator_precedence_17, "1 + (2 + 3) + 4");
debug_snapshot!(operator_precedence_19, "(5 + 5) * 2");
debug_snapshot!(operator_precedence_21, "2 / (5 + 5)");
debug_snapshot!(operator_precedence_23, "-(5 + 5)");
debug_snapshot!(operator_precedence_25, "!(true == true)");
debug_snapshot!(operator_precedence_27, "a + add(b * c) + d");
debug_snapshot!(operator_precedence_22, "add(a, 2 * 3, add(6, 7 * 8))");
debug_snapshot!(operator_precedence_24, "add(a + b + c * d / f + g)");
debug_snapshot!(operator_precedence_26, "a * [1, 2, 3, 4][b * c] * d");
debug_snapshot!(operator_precedence_28, "add(a * b[2], b[1], 2 * [1, 2][1])");

debug_snapshot!(if_expr_happy_1, "if (x) { x }");
debug_snapshot!(if_expr_happy_2, "if (x < y) { x }");
debug_snapshot!(if_expr_happy_3, "if (x < y) { x } else { y }");
debug_snapshot!(
    if_expr_happy_4,
    "if (x < y) { x } else { let z = x + y; z }"
);

debug_snapshot!(if_expr_unhappy_1, "if (x +) { x }");
debug_snapshot!(if_expr_unhappy_2, "if (x +) { x } else { x + 1 }");
debug_snapshot!(if_expr_unhappy_3, "if (x) { let x = 1; x < }");
debug_snapshot!(if_expr_unhappy_4, "if (x) { let x = 1");
debug_snapshot!(if_expr_unhappy_5, "if (x");
debug_snapshot!(if_expr_unhappy_6, "if (x { x }"); // NEEDS IMPROVEMENT
debug_snapshot!(if_expr_unhappy_7, "if (x) { x else { x }");
debug_snapshot!(if_expr_unhappy_8, "if (x) { x + } ");
debug_snapshot!(if_expr_unhappy_9, "if (x) { x + } else { x - 1; }");
debug_snapshot!(if_expr_unhappy_10, "if (@) { x }");
debug_snapshot!(if_expr_unhappy_11, "if (x) { @ }");
debug_snapshot!(if_expr_unhappy_12, "if (x) { x } else { @ }");

debug_snapshot!(fn_expr_happy_1, "fn() {}");
debug_snapshot!(fn_expr_happy_2, "fn(x) {}");
debug_snapshot!(fn_expr_happy_3, "fn(x, y, z) {}");
debug_snapshot!(fn_expr_happy_4, "fn(x) { return x; }");
debug_snapshot!(fn_expr_happy_5, "fn(x, y) { let z = x + y; return z; }");
debug_snapshot!(fn_expr_happy_6, "fn(x, y, z,) {}");
debug_snapshot!(fn_expr_happy_7, "fn(x, y, z,) { if (x) { y } else { z } }");

debug_snapshot!(fn_expr_unhappy_1, "fn( {}"); // NEEDS IMPROVEMENT
debug_snapshot!(fn_expr_unhappy_2, "fn(1+1) {}");
debug_snapshot!(fn_expr_unhappy_3, "fn(x y) {}");
debug_snapshot!(fn_expr_unhappy_4, "fn(x y z) {}");
debug_snapshot!(fn_expr_unhappy_5, "fn(x y, z) }{}");
debug_snapshot!(fn_expr_unhappy_6, "fn(, x) {}");
debug_snapshot!(
    fn_expr_unhappy_7,
    "fn(x, y, z,) { if (x) { y } else { z  }; 5"
); // NEEDS IMPROVEMENT
debug_snapshot!(fn_expr_unhappy_8, "fn(1+) {}");
debug_snapshot!(fn_expr_unhappy_9, "fn(x+) {}");
debug_snapshot!(fn_expr_unhappy_10, "fn(@) { return x; }");
debug_snapshot!(fn_expr_unhappy_11, "fn(@, x) { return x; }");
debug_snapshot!(fn_expr_unhappy_12, "fn(x) { @ }");
debug_snapshot!(fn_expr_unhappy_13, "fn(x) { x + 1; @ }");

debug_snapshot!(fn_call_happy_1, "add()");
debug_snapshot!(fn_call_happy_2, "add(x)");
debug_snapshot!(fn_call_happy_3, "add(x, y);");
debug_snapshot!(fn_call_happy_4, "fn(x, y){ return x + y; }(1, 2);");
debug_snapshot!(fn_call_happy_5, "add(x, y,);");
debug_snapshot!(fn_call_happy_6, "add(6, 7 * 8); 2");
debug_snapshot!(fn_call_happy_7, "add(2, add(6, 7 * 8))");

debug_snapshot!(fn_call_unhappy_1, "add(");
debug_snapshot!(fn_call_unhappy_2, "add)");
debug_snapshot!(fn_call_unhappy_3, "add(x y)");
debug_snapshot!(fn_call_unhappy_4, "add(x y, z)");
debug_snapshot!(fn_call_unhappy_5, "add(, x)");
debug_snapshot!(fn_call_unhappy_6, "add(x +)");
debug_snapshot!(fn_call_unhappy_7, "add(+)");
debug_snapshot!(fn_call_unhappy_8, "add(@)");
debug_snapshot!(fn_call_unhappy_9, "add(2, @)");
debug_snapshot!(fn_call_unhappy_10, "add(@, 2)");

debug_snapshot!(array_happy_1, "[1,2,3]; 5");
debug_snapshot!(array_happy_2, "[1,\"foo\",3];");
debug_snapshot!(array_happy_3, "[1,\"foo\",fn(x) { x + 1 }];");
debug_snapshot!(array_happy_4, "[1,2,3, [4]]; 5");

debug_snapshot!(array_unhappy_1, "[1,2 3]; 5");
debug_snapshot!(array_unhappy_2, "[1,2");
debug_snapshot!(array_unhappy_3, "[1,+]");
debug_snapshot!(array_unhappy_4, "[1,@]");

debug_snapshot!(hash_happy_1, "{}");
debug_snapshot!(hash_happy_2, r#"{"foo": "bar"}"#);
debug_snapshot!(hash_happy_3, r#"{"foo": "bar", "eggs": "spam"}"#);
debug_snapshot!(hash_happy_4, r#"{true: "bar", 2: "baz"}"#);
debug_snapshot!(hash_happy_5, r#"{"foo": "bar",}"#);

debug_snapshot!(hash_unhappy_1, "{A}");
debug_snapshot!(hash_unhappy_2, "{1:}");
debug_snapshot!(hash_unhappy_3, "{1:2 2:3}");
debug_snapshot!(hash_unhappy_4, "{1:2 2}");
debug_snapshot!(hash_unhappy_5, "{1:2");
debug_snapshot!(hash_unhappy_6, "{1:2; 5;");
debug_snapshot!(hash_unhappy_7, "{,1:2}");
debug_snapshot!(hash_unhappy_8, "{1:@}");
debug_snapshot!(hash_unhappy_9, "{@:2}");
debug_snapshot!(hash_unhappy_10, "{1:2, @:2}");

debug_snapshot!(array_index_happy_1, "foo[1]");
debug_snapshot!(array_index_happy_2, "foo[add(3)]");
debug_snapshot!(array_index_happy_3, "foo[1+2]");
debug_snapshot!(array_index_happy_4, "[1,2,3,4][1+2]");

debug_snapshot!(hash_index_happy_1, r#"{"foo": "bar"}["foo"]"#);
debug_snapshot!(hash_index_happy_2, r#"{true: "bar", 2: "baz"}[true]"#);
debug_snapshot!(hash_index_happy_3, r#"{true: "bar", 2: "baz"}[1+1]"#);

debug_snapshot!(array_index_unhappy_1, "foo[1");
debug_snapshot!(array_index_unhappy_2, "foo[1+");
debug_snapshot!(array_index_unhappy_3, "foo[1,");
debug_snapshot!(array_index_unhappy_4, "foo[1,2]");
debug_snapshot!(array_index_unhappy_5, "foo[@]");

debug_snapshot!(mixed, "let foo = 1;\n[1, 2];");
debug_snapshot!(mixed_err, "let foo;\n[1 2, 3];");

#[test]
fn errors() {
    let input = "
let x = 12;
let y;

let add = fn(x y) {
    return x y;
}

return y
x
";

    let program = Parser::from_source(input).parse_program();
    let errors = program
        .collect_errors()
        .into_iter()
        .map(|err| err.into())
        .collect::<Vec<SpannedDiagnostic>>();
    insta::with_settings!({
        description => input,
    }, {
        insta::assert_snapshot!(input_diagnostics(input, errors));
    })
}
