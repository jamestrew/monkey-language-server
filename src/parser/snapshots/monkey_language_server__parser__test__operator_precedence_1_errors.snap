---
source: src/parser/test.rs
description: "-a * b"
expression: "parser_errors(\"-a * b\", errors)"
---
-a * b
